### PR TITLE
Fix inference for generic in functor

### PR DESCRIPTION
**Pull request by OliverJAsh**
Without describing how the generic F is used in the interface, there's no way for TypeScript to perform inference when a functor is provided.

Example:

``` ts
{
    const functor = {
        map: (fn: (x: string) => string) => functor
    }
    S.map(x => x.trim(), functor) // Property 'trim' does not exist on type '{}'.
}
```

### CODE REVIEW

File: sanctuary/sanctuary-types.d.ts [hunk 1]
```diff
@@ -54,6 +54,7 @@ type NonZeroFiniteNumber = number; // - Infinity, -Infinity, 0, -0, NaN
 type ValidNumber = number; // - NaN
 
 export interface Functor<F> {
+  map<U>(fn: (f: F) => U): Functor<U>;
```
Comment (davidchambers) [inline sanctuary/sanctuary-types.d.ts]: Should this be `fantasy-land/map` rather than `map`, to match the method name? (I don't know TypeScript.)

Comment (OliverJAsh) [reply sanctuary/sanctuary-types.d.ts]: Ah, I didn't realise Sanctuary expects Fantasy Land methods to be prefixed with `fantasy-land/`? Is that part of Fantasy Land?

Comment (davidchambers) [reply sanctuary/sanctuary-types.d.ts]: Yes, it is. See [prefixed method names][1].


[1]: https://github.com/fantasyland/fantasy-land/tree/v3.1.0#prefixed-method-names

Comment (OliverJAsh) [reply sanctuary/sanctuary-types.d.ts]: Thanks @davidchambers, updated.

Comment (mnn) [reply sanctuary/sanctuary-types.d.ts]: But PR is against master, which is intended for Sanctuary 0.11. Weren't those prefixed methods added in 0.12?

Comment (OliverJAsh) [reply sanctuary/sanctuary-types.d.ts]: Updated base to `S_0.12` branch.

**Comment by OliverJAsh** [comment]
Updated base to `S_0.12` branch.

**Comment by mnn** [comment]
Sorry it took that long, but sadly I don't think I will be maintaining this typings anymore :confused:. (At work we are using a "methods" form, but that is being removed in new Sanctuary version. Function composition has issues with TypeScript [one has to manually specify a lot more types - inference is often not working], so using methods is, at this time, a much better choice. For now we are staying on the old version, but we will probably switch to something else [maybe I'll write something, we are not using that many things from Sanctuary], because only unique thing Sanctuary offers - dynamic type checking - is broken at that version).

**Comment by davidchambers** [comment]
@mnn, it would be quite easy to write a Sanctuary wrapper which modifies prototypes to support method chaining. For example:

```javascript
const S = require('sanctuary');

S.Maybe.prototype.map = S.Maybe.prototype['fantasy-land/map'];
...

module.exports = S;
```

**Comment by OliverJAsh** [comment]
You should check out fp-ts https://github.com/gcanti/fp-ts

**Pull request merged**